hail/src/test/scala/is/hail/expr/ir/IRSuite.scala

```diff
@@ -742,6 +740,929 @@ class IRSuite extends HailSuite {
     assertEvalsTo(MakeArray(FastSeq(), TArray(TInt32())), FastIndexedSeq())
   }
 
+  @Test def testMakeArrayInferPTypeFromNestedRef() {
+    var ir = MakeArray(FastSeq(I32(5), NA(TInt32()), I32(-3)), TArray(TInt32()))
```

Comment:
This test is still about two orders of magnitude too big. We've had problems with test maintainability in the past, so we need to be diligent about keeping tests legible and terse.

Part of the problem is that you're basically testing the type unification method here. You should make a separate test for that, where you write one assertion for each PType, and a couple of tests with nested types with structs and arrays to test integration. 

I think a good test of just the MakeArray inference (where _good_ means that it's clear what's being tested and how) would be:
 * define an env with ~3 different nested types
 * check inference of 0-arg, 1-arg, 2-arg MakeArrays with same and different types

Should be ~20 lines max.

Reply:
Got it, will fix tomorrow